jb/itch5/testing/ut_type_based_order_book.hpp

```diff
@@ -0,0 +1,331 @@
+  BOOST_CHECK_EQUAL(actual.second, 200);
+  // handler should return false
+  BOOST_CHECK_EQUAL(r, false);
+  // .. and the book_depth should be decremented
+  BOOST_CHECK_EQUAL(tested.count(), 2);
+
+  // ... deleting the best quote uncovers the best price ...
+  r = tested.reduce_order(price4_t(base_p + diff), 200);
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(base_p - diff));
+  BOOST_CHECK_EQUAL(actual.second, 300);
+  // handler should return true... it is an inside change
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should be decremented
+  BOOST_CHECK_EQUAL(tested.count(), 1);
+
+  // ... deleting the remaining price takes the book depth to 0
+  r = tested.reduce_order(price4_t(base_p - diff), 300);
+  actual = tested.best_quote();
+  auto emp = tested.empty_quote(); // gets empty_quoted already tested
+  BOOST_CHECK_EQUAL(actual.first, emp.first);
+  BOOST_CHECK_EQUAL(actual.second, emp.second);
+  // handler should return true... it is an inside change
+  BOOST_CHECK_EQUAL(r, true);
+  // .. and the book_depth should be decremented
+  BOOST_CHECK_EQUAL(tested.count(), 0);
+}
+
+/**
+*order_book type trivial test.
+*@tparam order_book type based order book to be tested
+*/
+template <typename order_book>
+void test_order_book_type_trivial(order_book& tested) {
+
+  typename order_book::config cfg;
+
+  typename order_book::buys_t buy_test(cfg);
+  typename order_book::sells_t sell_test(cfg);
+
+  test_side_type_trivial(buy_test);
+  test_side_type_trivial(sell_test);
+}
+
+/**
+ * order_book type error handling test.
+ * @tparam order_book type based order book to be tested
+ */
+template <typename order_book>
+void test_order_book_type_errors(order_book& tested) {
+  typename order_book::config cfg;
+
+  typename order_book::buys_t buy_test(cfg);
+  typename order_book::sells_t sell_test(cfg);
+
+  test_side_type_errors(buy_test);
+
+  test_side_type_errors(sell_test);
+}
+
+/**
+ * order_book type error handling array_based specific tests.
+ * @tparam order_book type based order book to be tested
+ */
+template <typename order_book>
+void test_order_book_type_errors_spec(order_book& tested) {
+  typename order_book::config cfg;
+
+  typename order_book::buys_t buy_test(cfg);
+  typename order_book::sells_t sell_test(cfg);
+
+  test_side_type_errors_spec(buy_test);
+
+  test_side_type_errors_spec(sell_test);
+}
+
+/**
+ * order_book type add and reduce handling test.
+ * @tparam order_book type based order book to be tested
+ */
+template <typename order_book>
+void test_order_book_type_add_reduce(order_book& tested) {
+  typename order_book::config cfg;
+
+  typename order_book::buys_t buy_test(cfg);
+  typename order_book::sells_t sell_test(cfg);
+
+  test_side_type_add_reduce(buy_test);
+
+  test_side_type_add_reduce(sell_test);
+}
+
+} // namespace testing
+} // namespace itch5
+} // namespace jb
+
+#endif // jb_itch5_ut_type_based_order_book_hpp
```

Comment:
ACK.

I am going to review what I got now, and reopen the same pull request....